File: practice/algorithms/sorting/insertion_sort-part_1/InsertionSort1.java
import java.util.*;

public class InsertionSort1 {
    
    public static void insertIntoSorted(int[] ar) {
        // Fill up this function
    	int num = ar[ar.length - 1];
    	boolean placed = false;
    	for(int j = ar.length - 2; j >= 0; j--)
    	{
    		if(ar[j] > num)
    		{
    			ar[j + 1] = ar[j];
    			printArray(ar);
    		}
    		else
    		{
    			ar[j + 1] = num;
    			printArray(ar);
    			placed = true;
    			break;
    		}
    	}
    	if(!placed)
    	{
    		ar[0] = num;
    		printArray(ar);
    	}
    }
    
    
/* Tail starts here */
     public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int s = in.nextInt();
        int[] ar = new int[s];
         for(int i=0;i<s;i++){
            ar[i]=in.nextInt(); 
         }
         insertIntoSorted(ar);
         in.close();
    }
    
    
    private static void printArray(int[] ar) {
      for(int n: ar){
         System.out.print(n+" ");
      }
        System.out.println("");
   }
    
    
}
 

File: practice/algorithms/sorting/insertion_sort-part_1/InsertionSort1.scala
import scala.io.Source
import scala.util.control.Breaks._

object InsertionSort1 extends App {
  val lines = Source.stdin.getLines.toList
  val arr = lines(1).split(" ").map(_.toInt)

  insertIntoSorted(arr)

  def insertIntoSorted(arr: Array[Int]): Unit = {
    val insertionValue = arr(arr.length - 1)
    var placed = false
    breakable {
      Range(arr.length - 2, -1, -1).foreach(i => {
        if (arr(i) > insertionValue) {
          arr(i + 1) = arr(i)
          printArray(arr)
        } else {
          arr(i + 1) = insertionValue
          printArray(arr)
          placed = true
          break
        }
      })
    }
    if(!placed) {
      arr(0) = insertionValue
      printArray(arr)
    }
  }

  def printArray(arr: Array[Int]): Unit = {
    println(arr.mkString(" "))
  }
}
